server/src/database/remoteStorage.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  S3Client,
+  ListBucketsCommand,
+  ListObjectsV2Command,
+  GetObjectCommand,
+  PutObjectCommand,
+  type ListBucketsCommandOutput,
+  type ListObjectsV2CommandOutput,
+  type GetObjectCommandOutput,
+  type PutObjectCommandOutput,
+  type S3ServiceException,
+} from '@aws-sdk/client-s3'
+import dotenv from 'dotenv'
+
+dotenv.config()
+
+export class StoreFile<T> {
+  private readonly directory!: string
+  private readonly storage!: IStorage
+
+  constructor (directory: string, storage: IStorage) {
+    this.directory = directory
+    this.storage = storage
+  }
+
+  getFile: (fileName: string) => Promise<T | null> = async (fileName: string) => await this.storage.getFile(this.directory, fileName)
+  putFile: (fileName: string, object: T) => Promise<void> = async (fileName: string, object: T) => { await this.storage.putFile(this.directory, fileName, object) }
+}
+
+interface IStorage {
+  listDirectories: () => Promise<string[]>
+  listFiles: (directory: string) => Promise<string[]>
+  getFile: <T>(directory: string, file: string) => Promise<T | null>
+  putFile: <T>(directory: string, file: string, object: T) => Promise<void>
+}
+
+export class RemoteStorage implements IStorage {
+  private readonly storage: S3Storage
+
+  constructor () {
+    this.storage = new S3Storage()
+  }
+
+  listDirectories = async (): Promise<string[]> => {
+    try {
+      const response = await this.storage.listBuckets()
+      if (response.$metadata.httpStatusCode === 200 && response.Buckets) {
+        return response.Buckets.map(({ Name }) => Name as string)
+      }
+    } catch (e) {
+      console.error(e)
+    }
+
+    return []
+  }
+
+  listFiles = async (directory: string): Promise<string[]> => {
+    try {
+      const response = await this.storage.listObjects(directory)
+      if (response.$metadata.httpStatusCode === 200 && response.Contents) {
+        return response.Contents.map(({ Key }) => Key as string)
+      }
+    } catch (e) {
+      console.error(e)
+    }
+
+    return []
+  }
+
+  getFile = async <T>(directory: string, file: string): Promise<T | null> => {
```

Comment:
I think this should EITHER:
A: just return the raw Buffer, and the conumer cna pass it to a parser
B: take in a parser that is responsible for creating the shape.  `JSON.parse(await response.Body.transformToString()) as T` is making a ton of assumptions